### ISSUE TITLE
RE datasets

### ISSUE THREAD

**Issue by luana-be** [dmis-lab/biobert#162]
Hello, 

I'm using the GAD and EUADR datasets for relation extraction and I'm noticing contradictory annotations in both sets.

Here is an example extracted from the EUADR test set:

```
15 The @GENE$ SNP could be considered as a genetic marker to predict the clinical course of patients suffering from oropharyngeal and @DISEASE$. 0
16 The @GENE$ SNP could be considered as a genetic marker to predict the clinical course of patients suffering from @DISEASE$ and hypopharyngeal cancer. 1
```

How are these sentences actually annotated? 

Thank you in advance,
Luana

**Comment by wangweifeng2018** [comment]
Same issue, have you got answer regarding this?

**Comment by luana-be** [comment]
> Same issue, have you got answer regarding this?

Not yet! I decided to do some cleaning up manually/using grep

**Comment by wonjininfo** [comment]
Hi all, 
Thank you for your interest in our paper. 

It seems like the examples could be categorized as annotation errors. 
The original paper https://doi.org/10.1016/j.jbi.2012.04.004 said that the dataset is annotated by experts.

For the given examples, they are from PMID : 18347176. https://pubmed.ncbi.nlm.nih.gov/18347176/
```
The T393C SNP could be considered as a genetic marker to predict the clinical course of patients suffering from oropharyngeal and hypopharyngeal cancer.
```
is from the end of the abstract.

When taking a look at the original EUADR corpus from https://biosemantics.erasmusmc.nl/index.php/resources/euadr-corpus , we can find an annotation file "18347176.txt" in it.

```
...
Target-Disorder	True	concept	T393C	1649	1654	annotator1,annotator2,annotator3	['sda/5', 'sda/2', 'sda/1']	27	SNP & Sequence variations
...
Target-Disorder	True	concept	oropharyngeal	1757	1770	annotator1,annotator2,annotator3	['sda/1', 'sda/10', 'sda/15']	28	Diseases & Disorders
Target-Disorder	True	concept	hypopharyngeal cancer	1775	1796	annotator1,Computer,annotator3	['umls/C0006581', 'sda/17']	29	Diseases & Disorders
...
Target-Disorder	False	relation	27	29	['sda/5', 'sda/2', 'sda/1']	['umls/C0006581', 'sda/17']	1649:1654	1775:1796	annotator1,annotator3	PA
Target-Disorder	True	relation	27	28	['sda/5', 'sda/2', 'sda/1']	['sda/1', 'sda/10', 'sda/15']	1649:1654	1757:1770	annotator1,annotator2,annotator3	PA
```
From the above lines, you can see the original annotated file says that T393C (id : 27 - from the 9th column) and oropharyngeal (28) have a relation (TRUE) but T393C (27) and hypopharyngeal cancer (29) do not (which seems to be an error to me as well). 

As a BioNLP researcher, I think RE datasets are difficult to make and sometimes contains erroneous examples as it requires an extensive manual work of healthcare professionals. 
Although we have a few wrong samples, I think providing a dataset is a huge contribution to the BioNLP community, and I would like to express my sincere gratitude to the annotators! 😊

Thank you!
Best regards,
Wonjin

------------
Here is a python script I used to check the dataset. 
```python
inptext=<get 18347176.txt>
inpTok = [ele.split("\t") for ele in inptext.splitlines()]
print(len(inpTok)) # should be about 52
entities = {int(ele[8]):"entity : '%s', "%ele[3]+ele[7] for ele in inpTok if ele[2]=="concept"}
print(len(entities)) # should be 36
entities[28] # > "entity : 'oropharyngeal', ['sda/1', 'sda/10', 'sda/15']"
```

**Comment by jambo6** [comment]
The dataset does not seem to be a few cases of poorly labelled samples, it seems to be extremely poor. Could something else have gone wrong somewhere?

For example, it fails on even the most obvious changes to the input sentence
```
# True sentence
sentence_true = 'This result suggests that susceptibility to @DISEASE$ may be associated with the RsaI and @GENE$ polymorphism of the P450IIE1 gene.'

# Negated sentence
sentence_negated = "This result suggests that susceptibility to @DISEASE$ is not associated with the RsaI and @GENE$ polymorphism of the P450IIE1 gene."

# Run the fine tuned model
nlp(sentence_true)    # Returns {'label': 'LABEL_0', 'score': 0.9942463040351868}
nlp(sentence_negated)   # Returns {'label': 'LABEL_0', 'score': 0.9936720728874207}
```
The model appears to predict almost zero difference in all the cases i've tried between a positive association and a negated version. This suggests to me it's not an artifact of just some poor labels.

**Comment by jambo6** [comment]
@luana-be if you did end up doing some manual processing, is there any chance you would share said labels?

**Comment by luana-be** [comment]
@jambo6 my manual processing was not enough to get good results. I totally gave up on these RE datasets! Sorry :-(

**Comment by jambo6** [comment]
Ah no worries. Thanks for the reply. 

Are you familiar with BioNLP by any chance? I know there exist good gene-gene relationship databases and I was wondering : is there anything that would likely prevent a model trained on gene-gene relationships perform poorly on gene-disease? It feels to me the language is _relatively_ similar. Obviously its not ideal and I'm sure you'd miss certain things but I wouldn't have thought it would be too bad...

**Comment by luana-be** [comment]
That's a good idea @jambo6 ! Thanks for the insight :-)

**Comment by amirkdv** [comment]
@wonjininfo thank you for looking into this.

I also ran into the exact same issue described here and documented my findings in #153. I agree with @jambo6 that this is a serious issue, beyond the usual "labeled data is hard". I'd even add that the scope goes beyond BioBert and this repository per se. At this point, this GAD RE dataset has become a _de facto_ benchmark, used by a lot of other folks, and BioBert is now part of the genealogy of the dataset (e.g. see BLURB).

I encourage you to look at my findings in #153; you can find a summary of it below.

---

I set up a tiny experiment. I picked 20 random examples from the [official BioBert GAD RE dataset](https://drive.google.com/file/d/1-jDKGcXREb2X9xTFnuiJ36PvsqoyHWcw/view?usp=drive_open) and verified their prescribed labels. I found that the true/false labels were basically **no better than a coin toss**. I then tried to trace the genealogy of the GAD RE dataset, which roughly goes something like this: [Becker et al. (2004)](https://www.nature.com/articles/ng0504-431), [Bravo et al. (2015)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4466840/), and Lee, Yoon, et al. 2019 (i.e. BioBert).

My conclusion so far is that the main problem with the datset we have now is the very definition of true/false labels. Becker et al. (2004)'s dataset was a good ole, manually curated RE dataset with 5,000 data point, each a tuple like `(pubmed, gene, disease, label)`, notably each sample refers to a PubMed id, i.e. whole article, and not the specific sentence in question. Then Bravo et al. (2015) do some non-trivial, and IMO questionable, gymnastics and turn the original GAD into a bigger (poorly) labeled RE dataset. This is where the labels get weird, and I think useless and BioBert's GAD RE dataset simply inherits the labels from Bravo et al. (2015). Specifically, it seems that a `(sentence, gene, disease)` would get a label of false not only if a human annotated that sentence as such (desired scenario), but also if the article from which it was taken was not part of the original GAD, for example if it was published after 2004!

It does sounds crazy, but it's my best theory of what's gone wrong and how badly.

**Comment by jambo6** [comment]
Yeah you are right on this, I just read _Gu et al., "Domain-Specific Language Model Pretraining for Biomedical
Natural Language Processing"_ and they state

_The Genetic Association Database corpus was created semi-automatically using the Genetic Association Archive. Specifically, the archive contains a list of gene-disease associations, with the corresponding
sentences in the PubMed abstracts reporting the association studies. Bravo et al. used a biomedical NER
tool to identify gene and disease mentions, and create the positive examples from the annotated sentences in the
archive, and negative examples from gene-disease co-occurrences that were not annotated in the archive._

Interestingly, this odd labelling method did not appear to be a cause for concern.

**Comment by DunnoHtL** [comment]
Thanks for your guys' exploration! I came across the same issue and was glad that I'm not the only one. By the way, does anyone know if there exists any high-quality dataset for RE tasks? Thanks ahead!

**Comment by jambo6** [comment]
I used this one recently and it worked quite well: https://github.com/sujunhao/RENET2

I stuck a fine-tuned model on huggingface if you are interested in trying it out: https://huggingface.co/jambo/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext-finetuned-renet

**Comment by DunnoHtL** [comment]
Appreciate your links so much! @jambo6, but the dataset link in that repo (http://www.bio8.cs.hku.hk/RENET2/renet2_data_models.tar.gz) doesn't seem to work :(

I also like your idea of utilizing gene-gene relationship databases. In fact, I did a little investigation and found DisGeNET may serve as a good resource. Take Alzheimer's Disease as an example, this page https://www.disgenet.org/browser/0/1/1/C0002395/ lists all the association types and the evidence (free text) related to it. I was just wondering whether we could constitute a corpus based on that and establish a multi-classification NLP task.

**Comment by DunnoHtL** [comment]
@jambo6 Thx I try IE and it works!

That's insane... Can't imagine this problematic dataset has been utilized by such a highly cited work...

Days ago a friend of mine just recommended me another source of biomedical NLP datasets: https://www.i2b2.org/NLP/DataSets/Main.php, you can also have a look.
